shards/shards.go

```diff
@@ -111,33 +111,38 @@ func (ss *shardedSearcher) Close() {
 	}
 }
 
-func selectRepoSet(shards []rankedShard, q query.Q) []rankedShard {
-	filtered := shards[:0]
+func selectRepoSet(shards []rankedShard, q query.Q) ([]rankedShard, query.Q) {
+	and, ok := q.(*query.And)
+	if !ok {
+		return shards, q
+	}
 
-	eval := query.Map(q, func(q query.Q) query.Q {
-		setQuery, ok := q.(*query.RepoSet)
+	for i, c := range and.Children {
+		setQuery, ok := c.(*query.RepoSet)
 		if !ok {
-			return q
+			continue
 		}
 
+		filtered := shards[:0]
+
 		for _, s := range shards {
-			if repositorer, ok := s.Searcher.(Repositorer); ok {
+			if repositorer, ok := s.Searcher.(repositorer); ok {
 				repo := repositorer.Repository()
 				if setQuery.Set[repo.Name] {
 					filtered = append(filtered, s)
 				}
 			}
 		}
+		and.Children[i] = &query.Const{Value: len(filtered) > 0}
 
-		return &query.Const{Value: true}
-	})
-	query.Simplify(eval)
-
-	if len(filtered) != 0 {
-		return filtered
+		// Stop after first RepoSet, otherwise we might append duplicate
+		// shards to `filtered`
+		if len(filtered) != 0 {
+			return filtered, and
```

Comment:
I think you can just always return here, don't have to check len of filtered.

Also return `query.Simplify(and)`, to make Simplify do nothing per shard.